finality-aleph/src/testing/data_io.rs

```diff
@@ -0,0 +1,156 @@
+use crate::data_io::{AlephNetworkMessage, DataStore};
+use futures::{
+    channel::{
+        mpsc::{self, UnboundedReceiver, UnboundedSender},
+        oneshot,
+    },
+    StreamExt,
+};
+use futures_timer::Delay;
+use sc_block_builder::BlockBuilderProvider;
+use sp_consensus::BlockOrigin;
+use std::{future::Future, sync::Arc, time::Duration};
+use substrate_test_runtime_client::{
+    runtime::{Block, Hash},
+    Backend, ClientBlockImportExt, DefaultTestClientBuilderExt, TestClient, TestClientBuilder,
+    TestClientBuilderExt,
+};
+
+#[derive(Debug)]
+struct TestNetworkData {
+    data: Vec<Hash>,
+}
+
+impl AlephNetworkMessage<Block> for TestNetworkData {
+    fn included_blocks(&self) -> Vec<Hash> {
+        self.data.to_vec()
+    }
+}
+
+fn prepare_data_store() -> (
+    impl Future<Output = ()>,
+    Arc<TestClient>,
+    UnboundedSender<TestNetworkData>,
+    UnboundedReceiver<TestNetworkData>,
+    oneshot::Sender<()>,
+) {
+    let client = Arc::new(TestClientBuilder::new().build());
+
+    let (aleph_network_tx, data_store_rx) = mpsc::unbounded();
+    let (data_store_tx, aleph_network_rx) = mpsc::unbounded();
+    let mut data_store = DataStore::<Block, TestClient, Backend, TestNetworkData>::new(
+        client.clone(),
+        data_store_tx,
+        data_store_rx,
+    );
+    let (exit_data_store_tx, exit_data_store_rx) = oneshot::channel();
+
+    (
+        async move {
+            data_store.run(exit_data_store_rx).await;
+        },
+        client,
+        aleph_network_tx,
+        aleph_network_rx,
+        exit_data_store_tx,
+    )
+}
+
+async fn import_blocks(client: &mut Arc<TestClient>, n: u32) -> Vec<Hash> {
+    let mut blocks = vec![];
+    for _ in 0..n {
+        let block = client
+            .new_block(Default::default())
+            .unwrap()
+            .build()
+            .unwrap()
+            .block;
+        client
+            .import(BlockOrigin::Own, block.clone())
+            .await
+            .unwrap();
+        blocks.push(block.header.hash());
+    }
+    blocks
+}
+
+#[tokio::test]
+async fn sends_messages_with_imported_blocks() {
+    let (task_handle, mut client, store_tx, mut store_rx, exit_data_store_tx) =
+        prepare_data_store();
+
+    let consensus_network_handle = tokio::spawn(task_handle);
+
+    let blocks = import_blocks(&mut client, 4).await;
+
+    store_tx
+        .unbounded_send(TestNetworkData {
+            data: blocks.clone(),
+        })
+        .unwrap();
+
+    Delay::new(Duration::from_millis(10)).await;
+    exit_data_store_tx.send(()).unwrap();
+    consensus_network_handle.await.unwrap();
+    let message = store_rx.next().await.expect("We own the tx");
```

Comment:
Can you shuffle the 4 lines above so that delay is not needed? 
1. `store_rx.next().await`
2. `exit_data_store_tx.send(())`
3. `consensus_network_handle.await`

Reply:
I wanted this test to not run infinitely in case the DataStore wouldn't send the message. The delay is needed here so that DataStore would receive message before exit signal. If we send the exit signal after waiting for `store_rx.next()` it is possible that if DataStore has a bug this test will run forever.

Reply:
If there was a bug, you would see it after 60s, i.e. `cargo test` would output a warning. I wouldn't add sleeps for such a case.

Reply:
Yes, there is a warning, but it doesn't stop the test, so I'm not sure how long it would run for example in our pipeline on github. The delay is here just so that `unbounded_channel` is handled before `one_shot`. I could also add here a limit, so that after some time it panics (for example 30 seconds).

Reply:
I'm not sure what are the limits on GH but hanging test may happen in other situations as well and we don't add artificial panics to dodge them.

Reply:
Ok, then I will remove the delay